README
Games-Ratings version 0.0.4

This is Games::Ratings, a module to calculate ratings for chess or
other games. At the moment one can calculate two different kinds of
chess ratings: FIDE ratings (aka "FIDE Elo") and German DWZ.
Furthermore calculation of go ratings (EGF) is provided. For more
informations see POD of:
* Games::Ratings
* Games::Ratings::Chess::FIDE
* Games::Ratings::Chess::DWZ
* Games::Ratings::Go::EGF

Warning!

This is the first module I wrote. Certainly it can be improved in
various ways. Use at your own risk.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install


TODO

* *.pm:            check arguments for subroutines
                   maybe 'use Params::Validate'
* *.pm:            Error handling (croak() etc.)
* *.pm:            complete documentation (error messages etc.)
* *.pm:            write more tests
* FIDE.pm:         rating for previously unrated player
* DWZ.pm:          rating for previously unrated player
* DWZ.pm:          what to do if opponent is without DWZ rating?
* EGF.pm:          rating for previously unrated player


DEPENDENCIES

This module needs Perl 5.6 or higher.

Also, this module depends on:

  'Test::More'        => 0
  'Class::Std::Utils' => 0
  'version'           => 0


COPYRIGHT AND LICENCE

Copyright (C) 2008, Christian Bartolomaeus

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Games-Ratings

0.0.1  Wed Feb 13 14:51:49 2008
       Initial release.

0.0.2  Thu Feb 28 10:22:00 2008
       added 'use 5.6.1;' to Ratings.pm, DWZ.pm and FIDE.pm
         (those modules use 'our' which was introduced in Perl 5.6)
       added method $player->get_points_expected() to FIDE.pm
         (calculate expected points for given games)
       added method $player->get_performance() to FIDE.pm
         (calculate performance for given games)

0.0.3  Fri Feb 29 17:39:00 2008
       added 'use 5.006;' to Makefile.PL
         (I still got failed tests from cpan-testers for Perl 5.5)
       added Games::Ratings::Go::EGF (version 0.0.1)
         (calculation of go ratings)

0.0.4  Thu Jul 3  15:00:00 2008
       updated links to FIDE Handbook (FIDE changed their URLs)
       fixed small bug in Ratings.pm
         (subroutine get_number_of_games_played had no explicit return scalar)
	 (it has now ;-)


lib/Games/Ratings.pm
package Games::Ratings;

## TODO: Error handling
##       * croak()
##       check arguments for subroutines (use Data::Checker)?

use warnings;
use strict;
use Carp;

use 5.6.1;               # 'our' was introduced in perl 5.6
use version; our $VERSION = qv('0.0.4');

use Class::Std::Utils;   # we are using inside-out objects

{
    ## objects of this class (players) will have the following attributes
    my %rating_of;       # rating of player
    my %coefficient_of;  # coefficient of player
    my %games_of;        # list of games of player

    ## create new object (inside-out object -- see "Encapsulation" in PBP)
    sub new {
        my ($class) = @_;

        ## bless a scalar to instantiate the new object
        my $new_player_object = bless anon_scalar(), $class;

        return $new_player_object;
    }

    ## set rating according to passed argument
    sub set_rating {
        my ($self, $rating) = @_;

        ## check that rating is passed as argument
        croak( 'Usage: $obj->set_rating($rating)' )
            if @_ < 2;

        ## store rating in player object
        $rating_of{ident $self} = $rating;

        return;
    }
    
    ## return previously set rating
    sub get_rating {
        my ($self) = @_;
        return $rating_of{ident $self};
    }
    
    ## set coefficient according to passed argument
    sub set_coefficient {
        my ($self, $coefficient) = @_;

        ## check that coefficient is passed as argument
        croak( 'Usage: $obj->set_coefficient($coefficient)' )
            if @_ < 2;
        
        ## store coefficient in player object
        $coefficient_of{ident $self} = $coefficient;

        return;
    }
    
    ## return previously set coefficient
    sub get_coefficient {
        my ($self) = @_;
        return $coefficient_of{ident $self};
    }
    
    ## add new game to list of games ($games_of{ident $self} is list of games)
    sub add_game {
        my ($self, $game_ref) = @_;

        ## check that new game is passed as argument (hash reference) 
        croak( 'Usage: $obj->add_game( { 
                         opponent_rating => 2300,
                         result          => \'draw\',
                       }
                     );
                     ' )
            if @_ < 2;

        ## store additional game in player object
        push @{ $games_of{ident $self} }, $game_ref;

        return;
    }

    ## return all previously added games as a list of hash references
    sub get_all_games {
        my ($self) = @_;

        ## return list of games or throw an error if $self doesn't have games
        if ( $games_of{ident $self} ) {
            return @{ $games_of{ident $self} };
        }
        else {
            croak( 'There aren\'t any games played. Please use add_game().' )
        }
    }
    
    ## delete all previously added games
    sub remove_all_games {
        my ($self) = @_;
        $games_of{ident $self} = undef;
        return;
    }

    ## clean up attributes when object is destroyed
    sub DESTROY {
        my ($self) = @_;

        delete $rating_of{ident $self};
        delete $coefficient_of{ident $self};
        delete $games_of{ident $self};

        return;
    }
}
    
## return number of played games
sub get_number_of_games_played {
    my ($self) = @_;

    ## number of played games equals length of array of played games
    return scalar $self->get_all_games();
}

## calculate and return scored points
sub get_points_scored {
    my ($self) = @_;
    
    ## compute scored points from list of played games
    my $points_scored;
    foreach my $game_ref ( $self->get_all_games() ) {
        $points_scored += _get_numerical_result( $game_ref->{result} );
    }

    ## return scored points
    return $points_scored;
}

## calculate percentage score
sub get_percentage_score {
    my ($self) = @_;

    ## compute percentage score
    my $percentage_score = $self->get_points_scored() 
                           / $self->get_number_of_games_played();

    ## return percentage score
    return $percentage_score;
}

## calculate and return return average rating of opponents
sub get_average_rating_of_opponents {
    my ($self) = @_;
    
    ## calculate average rating of opponents from (list of) stored games
    my $rat_opps;
    foreach my $game_ref ( $self->get_all_games() ) {
        $rat_opps += $game_ref->{opponent_rating},
    }
    $rat_opps = sprintf( "%.f", $rat_opps 
                                / $self->get_number_of_games_played() );

    ## return average rating of opponents
    return $rat_opps;
}

## define lookup table for conversion 'verbal results' -> 'numerical results'
my %numerical_results = (
    win  =>   1,
    draw => 0.5,
    loss =>   0,
);

## get numerical result of $result (win, draw, loss)
sub _get_numerical_result {
    my ($result) = @_;

    ## numerical result is looked up in a small table (see above)
    my $numerical_result = $numerical_results{$result};

    ## return numerical result
    return $numerical_result;
}

1; # Magic true value required at end of module
__END__


=head1 NAME

Games::Ratings - generic methods for rating calculation (e.g. chess ratings)


=head1 VERSION

This document describes Games::Ratings version 0.0.4


=head1 SYNOPSIS

 ## not very useful, but one could do
 
 use Games::Ratings;
 my $player = Games::Ratings->new();
 $player->set_rating(2240);
 $player->add_game( {
                      opponent_rating => 2114,
                      result          => 'win',   ## or 'draw' or 'loss'
                    }
                  );
 my $n                = $player->get_number_of_games_played();
 my $points_scored    = $player->get_points_scored();
 my $opponents_rating = $player->get_average_rating_of_opponents();
 $player->remove_all_games;

  
=head1 DESCRIPTION

Games::Ratings provides some generic methods for other, more specific modules
(like Games::Ratings::Chess::FIDE or Games::Ratings::Chess::DWZ). It isn't
very useful for itself.


=head1 INTERFACE 

The following methods can be accessed from more specific modules -- like
Games::Ratings::Chess::FIDE. Also specific methods for rating calculation are
defined in those other modules.

=head2 new

  my $player = Games::Rating->new();

Create an object for storing rating and other data (like the FIDE development
coefficient) for one player as well as informations about rated games of this
player against some opponents.

=head2 set_rating

  $player->set_rating(2235);

Set rating of player.

=head2 get_rating

  my $own_rating = $player->get_rating();

Get rating of player.

=head2 set_coefficient

  $player->set_coefficient(15);

Set development coefficient of player. (A development coefficient is needed
for calculation of some ratings:
 * FIDE Elo (FIDE Handbook, B.02.10.52) 
 * German DWZ (Wertungsordnung des DSB, Punkt 4.9.2)).

=head2 get_coefficient

  my $own_coefficient = $player->get_coefficient();

Get development coefficient of player.

=head2 add_game

  $player->add_game( {
                       opponent_rating => 2184,
                       result          => 'draw',
                     }
                   );

Add rated game to object. We need opponents rating and the result of the game.
Results are one of 'win', 'draw' or 'loss'. The data is passed as hash
reference.

=head2 get_all_games

  my @list_of_games = $player->get_all_games();

Get list of previously added games. Each game is represented by a hash
reference (cmp. $player->add_game() ).

=head2 remove_all_games

  $player->remove_all_games();

Remove all games added (via 'add_game') so far.

=head2 DESTROY

  $player->DESTROY();

Remove object $player.

=head2 get_number_of_games_played

  my $n = $player->get_number_of_games_player();

Get number of games played.

=head2 get_points_scored

  my $points_scored = $player->get_points_scored();

Get total points scored in played games. A win gives 1 point, a draw gives 0.5
points, a loss gives 0 points.

=head2 get_percentage_score

  my $percentage_score = $player->get_percentage_score();

Get percentage score in played games.

=head2 get_average_rating_of_opponents

  my $opponents_rating = $player->get_average_rating_of_opponents();

Get average rating of opponents for played games.


=head1 DIAGNOSTICS
 
At the moment, there are no error or warning messages that the module can
generate.

=for author to fill in:
    List every single error and warning message that the module can
    generate (even the ones that will "never happen"), with a full
    explanation of each problem, one or more likely causes, and any
    suggested remedies.

=over

=item C<< Error message here, perhaps with %s placeholders >>

[Description of error here]

=item C<< Another error message here >>

[Description of error here]

=back

 
=head1 CONFIGURATION AND ENVIRONMENT

Games::Ratings requires no configuration files or environment variables.


=head1 DEPENDENCIES

Requires the C<Class::Std::Utils> module and the C<version> module. Needs Perl
5.6.1 or higher.


=head1 INCOMPATIBILITIES

None reported.


=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-games-ratings@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Christian Bartolomaeus  C<< <bartolin@gmx.de> >>


=head1 ACKNOWLEDGMENTS

This module was inspired by Terrence Brannon's module Chess::Elo
(http://search.cpan.org/~tbone/Chess-Elo/).


=head1 SEE ALSO

http://en.wikipedia.org/wiki/Elo_rating for informations about the Elo system.


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, Christian Bartolomaeus C<< <bartolin@gmx.de> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Games/Ratings/Chess/FIDE.pm
package Games::Ratings::Chess::FIDE;

## TODO: check arguments for subroutines (use Data::Checker)?
## TODO: Error handling
##       * croak()
##       * perldoc anpassen

use strict;
use warnings;
use Carp;

use 5.6.1;               # 'our' was introduced in perl 5.6
use version; our $VERSION = qv('0.0.4');

## look in Games::Ratings for methods not provide by this package
use base qw ( Games::Ratings );

## calculate rating change
sub get_rating_change {
    my ($self) = @_;

    ## get own rating and own coefficient
    my $own_rating      = $self->get_rating();
    my $own_coefficient = $self->get_coefficient();

    my $rating_change_total;
    ## calculate rating change for each game separately
    foreach my $game_ref ( $self->get_all_games() ) {
        ## add rating change for single game to total rating change
        $rating_change_total += _calc_rating_change_for_single_game(
                                    $own_rating,
                                    $own_coefficient,
                                    $game_ref->{opponent_rating},
                                    $game_ref->{result},
                                );
    }

    ## return total rating change
    return $rating_change_total;
}

## calculate new rating
sub get_new_rating {
    my ($self) = @_;

    ## $R_o -- old rating
    my $R_o = $self->get_rating();

    ## $R_n -- new rating (rounded)
    my $R_n = sprintf( "%.f", $R_o + $self->get_rating_change() );

    ## return new rating
    return $R_n;
}

## calculate expected points
sub get_points_expected {
    my ($self) = @_;
    
    ## $W_e -- expected points
    my $W_e;

    ## $own_rating -- own rating
    my $own_rating = $self->get_rating();

    ## sum up expected points for all games
    foreach my $game_ref ( $self->get_all_games() ) {
        $W_e += _get_scoring_probability_for_single_game(
                    $own_rating,
                    $game_ref->{opponent_rating},
                );
    }

    ## return expected points
    return $W_e;
}

## calculate performance
sub get_performance {
    my ($self) = @_;

    ## $R_h -- performance (independent from old rating)
    my $R_h;

    ## average rating of opponents
    my $R_c = $self->get_average_rating_of_opponents();

    ## $P -- percentage score (two digits needed)
    my $P = sprintf( "%.2f", $self->get_percentage_score() );

    ## if player scored 100 % or 0 % it's not possible to calc. performance
    if ($P == 1) {
        $R_h = $R_c + 667;
        return $R_h;
    }
    if ($P == 0) {
        $R_h = $R_c - 667;
        return $R_h;
    }

    ## lookup $D rating difference according to $P from probability table 
    my $D = _get_rating_difference_matching_percentage_score($P);

    ## calculate performance
    $R_h = $R_c + $D;

    ## return performance
    return $R_h;
}

########################
## internal functions ##
########################

## scoring probabilities depending from rating difference (FIDE B0210.1b)
our %scoring_probability_lookup_table;
_set_scoring_probability_lookup_table();

## lookup table needed to determine performance (FIDE B0210.1a)
our %reverse_scoring_probability_lookup_table;
_set_reverse_scoring_probability_lookup_table();

## calculate rating change for single game
sub _calc_rating_change_for_single_game {
    my ($A_rating, $A_coefficient, $B_rating, $result) = @_;
  
    ## get numerical result ( win=>1 draw=>0.5 loss=>0 )
    my $numerical_result = Games::Ratings::_get_numerical_result($result);

    ## check whether development coefficient is provided -- guess otherwise
    if (! defined $A_coefficient) {
        $A_coefficient = _guess_coefficient($A_rating);
    }

    ## get scoring probability for player A
    my $A_exp = _get_scoring_probability_for_single_game($A_rating,$B_rating);

    ## compute rating changes for player A
    my $A_rating_change = $A_coefficient * ($numerical_result-$A_exp);
  
    ## return rating changes for player A
    return ($A_rating_change);
}

## try to guess development coefficient
sub _guess_coefficient {
    my ($player_rating) = @_;

    ## guess coefficient according to rating (cmp. FIDE handbook B0210.52)
    my $player_coefficient;
    if ($player_rating >= 2400) {
        $player_coefficient = 10;
    }
    else {
        $player_coefficient = 15;
    }

    ## return guessed coefficient
    return $player_coefficient;
}

## calculate scoring probability for a single game
sub _get_scoring_probability_for_single_game {
    my ($A_rating,$B_rating) = @_;

    my $rating_difference = _get_rating_difference($A_rating,$B_rating);

    ## get scoring probability of player A from lookup table 
    my $A_exp;
    if ($rating_difference >= 0) {
        $A_exp = $scoring_probability_lookup_table{$rating_difference};
    }
    else {
        $A_exp = 1 - $scoring_probability_lookup_table{0-$rating_difference};
    }

    ## return scoring probability for player A
    return ($A_exp);
}

## calculate rating difference which is used to calc the scoring probability
sub _get_rating_difference {
    my ($A, $B) = @_;

    ## compute real rating difference
    my $rating_difference = ( $A-$B );

    ## large rating differences are cut (FIDE handbook B0210.51, 2nd sentence)
    if ($rating_difference > '350') {
        $rating_difference = '350';
    }
    if ($rating_difference < '-350') {
        $rating_difference = '-350';
    }

    ## return rating difference used for rating calculations
    return $rating_difference;
}

## calculate rating differences matching percentage score
sub _get_rating_difference_matching_percentage_score {
    my ($P) = @_;

    ## lookup $D (rating difference) from lookup table
    my $D;
    if ($P lt 0.5) {
        $D = -($reverse_scoring_probability_lookup_table{1-$P});
    }
    else {
        $D = $reverse_scoring_probability_lookup_table{$P};
    }

    ## return $D
    return $D;
}

## use hash as lookup table for scoring probability (cmp. FIDE B0210.1b)
sub _set_scoring_probability_lookup_table {
    foreach my $rating_difference (0..3) {
        $scoring_probability_lookup_table{$rating_difference} = 0.50;
    }
    foreach my $rating_difference (4..10) {
        $scoring_probability_lookup_table{$rating_difference} = 0.51;
    }
    foreach my $rating_difference (11..17) {
        $scoring_probability_lookup_table{$rating_difference} = 0.52;
    }
    foreach my $rating_difference (18..25) {
        $scoring_probability_lookup_table{$rating_difference} = 0.53;
    }
    foreach my $rating_difference (26..32) {
        $scoring_probability_lookup_table{$rating_difference} = 0.54;
    }
    foreach my $rating_difference (33..39) {
        $scoring_probability_lookup_table{$rating_difference} = 0.55;
    }
    foreach my $rating_difference (40..46) {
        $scoring_probability_lookup_table{$rating_difference} = 0.56;
    }
    foreach my $rating_difference (47..53) {
        $scoring_probability_lookup_table{$rating_difference} = 0.57;
    }
    foreach my $rating_difference (54..61) {
        $scoring_probability_lookup_table{$rating_difference} = 0.58;
    }
    foreach my $rating_difference (62..68) {
        $scoring_probability_lookup_table{$rating_difference} = 0.59;
    }
    foreach my $rating_difference (69..76) {
        $scoring_probability_lookup_table{$rating_difference} = 0.60;
    }
    foreach my $rating_difference (77..83) {
        $scoring_probability_lookup_table{$rating_difference} = 0.61;
    }
    foreach my $rating_difference (84..91) {
        $scoring_probability_lookup_table{$rating_difference} = 0.62;
    }
    foreach my $rating_difference (92..98) {
        $scoring_probability_lookup_table{$rating_difference} = 0.63;
    }
    foreach my $rating_difference (99..106) {
        $scoring_probability_lookup_table{$rating_difference} = 0.64;
    }
    foreach my $rating_difference (107..113) {
        $scoring_probability_lookup_table{$rating_difference} = 0.65;
    }
    foreach my $rating_difference (114..121) {
        $scoring_probability_lookup_table{$rating_difference} = 0.66;
    }
    foreach my $rating_difference (122..129) {
        $scoring_probability_lookup_table{$rating_difference} = 0.67;
    }
    foreach my $rating_difference (130..137) {
        $scoring_probability_lookup_table{$rating_difference} = 0.68;
    }
    foreach my $rating_difference (138..145) {
        $scoring_probability_lookup_table{$rating_difference} = 0.69;
    }
    foreach my $rating_difference (146..153) {
        $scoring_probability_lookup_table{$rating_difference} = 0.70;
    }
    foreach my $rating_difference (154..162) {
        $scoring_probability_lookup_table{$rating_difference} = 0.71;
    }
    foreach my $rating_difference (163..170) {
        $scoring_probability_lookup_table{$rating_difference} = 0.72;
    }
    foreach my $rating_difference (171..179) {
        $scoring_probability_lookup_table{$rating_difference} = 0.73;
    }
    foreach my $rating_difference (180..188) {
        $scoring_probability_lookup_table{$rating_difference} = 0.74;
    }
    foreach my $rating_difference (189..197) {
        $scoring_probability_lookup_table{$rating_difference} = 0.75;
    }
    foreach my $rating_difference (198..206) {
        $scoring_probability_lookup_table{$rating_difference} = 0.76;
    }
    foreach my $rating_difference (207..215) {
        $scoring_probability_lookup_table{$rating_difference} = 0.77;
    }
    foreach my $rating_difference (216..225) {
        $scoring_probability_lookup_table{$rating_difference} = 0.78;
    }
    foreach my $rating_difference (226..235) {
        $scoring_probability_lookup_table{$rating_difference} = 0.79;
    }
    foreach my $rating_difference (236..245) {
        $scoring_probability_lookup_table{$rating_difference} = 0.80;
    }
    foreach my $rating_difference (246..256) {
        $scoring_probability_lookup_table{$rating_difference} = 0.81;
    }
    foreach my $rating_difference (257..267) {
        $scoring_probability_lookup_table{$rating_difference} = 0.82;
    }
    foreach my $rating_difference (268..278) {
        $scoring_probability_lookup_table{$rating_difference} = 0.83;
    }
    foreach my $rating_difference (279..290) {
        $scoring_probability_lookup_table{$rating_difference} = 0.84;
    }
    foreach my $rating_difference (291..302) {
        $scoring_probability_lookup_table{$rating_difference} = 0.85;
    }
    foreach my $rating_difference (303..315) {
        $scoring_probability_lookup_table{$rating_difference} = 0.86;
    }
    foreach my $rating_difference (316..328) {
        $scoring_probability_lookup_table{$rating_difference} = 0.87;
    }
    foreach my $rating_difference (329..344) {
        $scoring_probability_lookup_table{$rating_difference} = 0.88;
    }
    foreach my $rating_difference (345..350) {
        $scoring_probability_lookup_table{$rating_difference} = 0.89;
    }
}

## use hash as lookup table (rating differences given a percentage score)
## (cmp. FIDE B0210.1a)
sub _set_reverse_scoring_probability_lookup_table {
    $reverse_scoring_probability_lookup_table{'0.50'} = '0';
    $reverse_scoring_probability_lookup_table{'0.51'} = '7';
    $reverse_scoring_probability_lookup_table{'0.52'} = '14';
    $reverse_scoring_probability_lookup_table{'0.53'} = '21';
    $reverse_scoring_probability_lookup_table{'0.54'} = '29';
    $reverse_scoring_probability_lookup_table{'0.55'} = '36';
    $reverse_scoring_probability_lookup_table{'0.56'} = '43';
    $reverse_scoring_probability_lookup_table{'0.57'} = '50';
    $reverse_scoring_probability_lookup_table{'0.58'} = '57';
    $reverse_scoring_probability_lookup_table{'0.59'} = '65';
    $reverse_scoring_probability_lookup_table{'0.60'} = '72';
    $reverse_scoring_probability_lookup_table{'0.61'} = '80';
    $reverse_scoring_probability_lookup_table{'0.62'} = '87';
    $reverse_scoring_probability_lookup_table{'0.63'} = '95';
    $reverse_scoring_probability_lookup_table{'0.64'} = '102';
    $reverse_scoring_probability_lookup_table{'0.65'} = '110';
    $reverse_scoring_probability_lookup_table{'0.66'} = '117';
    $reverse_scoring_probability_lookup_table{'0.67'} = '125';
    $reverse_scoring_probability_lookup_table{'0.68'} = '133';
    $reverse_scoring_probability_lookup_table{'0.69'} = '141';
    $reverse_scoring_probability_lookup_table{'0.70'} = '149';
    $reverse_scoring_probability_lookup_table{'0.71'} = '158';
    $reverse_scoring_probability_lookup_table{'0.72'} = '166';
    $reverse_scoring_probability_lookup_table{'0.73'} = '175';
    $reverse_scoring_probability_lookup_table{'0.74'} = '184';
    $reverse_scoring_probability_lookup_table{'0.75'} = '193';
    $reverse_scoring_probability_lookup_table{'0.76'} = '202';
    $reverse_scoring_probability_lookup_table{'0.77'} = '211';
    $reverse_scoring_probability_lookup_table{'0.78'} = '220';
    $reverse_scoring_probability_lookup_table{'0.79'} = '230';
    $reverse_scoring_probability_lookup_table{'0.80'} = '240';
    $reverse_scoring_probability_lookup_table{'0.81'} = '251';
    $reverse_scoring_probability_lookup_table{'0.82'} = '262';
    $reverse_scoring_probability_lookup_table{'0.83'} = '273';
    $reverse_scoring_probability_lookup_table{'0.84'} = '284';
    $reverse_scoring_probability_lookup_table{'0.85'} = '296';
    $reverse_scoring_probability_lookup_table{'0.86'} = '309';
    $reverse_scoring_probability_lookup_table{'0.87'} = '322';
    $reverse_scoring_probability_lookup_table{'0.88'} = '336';
    $reverse_scoring_probability_lookup_table{'0.89'} = '351';
    $reverse_scoring_probability_lookup_table{'0.90'} = '366';
    $reverse_scoring_probability_lookup_table{'0.91'} = '383';
    $reverse_scoring_probability_lookup_table{'0.92'} = '401';
    $reverse_scoring_probability_lookup_table{'0.93'} = '422';
    $reverse_scoring_probability_lookup_table{'0.94'} = '444';
    $reverse_scoring_probability_lookup_table{'0.95'} = '470';
    $reverse_scoring_probability_lookup_table{'0.96'} = '501';
    $reverse_scoring_probability_lookup_table{'0.97'} = '538';
    $reverse_scoring_probability_lookup_table{'0.98'} = '589';
    $reverse_scoring_probability_lookup_table{'0.99'} = '677';
}


1; # Magic true value required at end of module
__END__


=head1 NAME
 
Games::Ratings::Chess::FIDE - calculate changes to FIDE ratings (Elos)
 

=head1 VERSION
 
This document describes Games::Ratings::Chess::FIDE version 0.0.4
 

=head1 SYNOPSIS
 
 use Games::Ratings::Chess::FIDE;

 my $player = Games::Ratings::Chess::FIDE->new();
 $player->set_rating(2240);
 $player->set_coefficient(15);
 $player->add_game( {
                      opponent_rating => 2114,
                      result          => 'win',   ## or 'draw' or 'loss'
                    }
                  );

 my $rating_change = sprintf( "%+.2f", $player->get_rating_change() );
 my $new_rating = $player->get_new_rating();


=head1 DESCRIPTION

This module provides methods to calculate FIDE rating (Elo) changes for one
player, having played one or more rated games. Gains and losses are calculated
according to the FIDE rating rules (cmp. FIDE Rating Regulations, FIDE
Handbook B.02.10: http://www.fide.com/info/handbook?id=75&view=article).

FIDE uses a table with scoring probabilities in dependence from rating
differences between the players.

FIDE does _not_ use the formula P = 1/(1 + 10 ^ [D/400]).

Furthermore FIDE uses a development coefficient (K) depending on the current
rating and the number of rated games.
 * K = 25 for a player with a total of less than 30 games.
 * K = 15 as long as a player`s rating remains under 2400.
 * K = 10 once a player`s published rating has reached 2400
 

=head1 INTERFACE 

This modules provides the following methods specific to FIDE ratings. Other
(more generic) methods for rating calculation are provided by Games::Ratings.
Please check the documentation of Games::Ratings for those methods.

=head2 get_rating_change

  my $rating_change = sprintf("%+.2f", $player->get_rating_change() );

Calculate rating changes for all stored games and return sum of those
changes.

=head2 get_new_rating

  my $new_rating = $player->get_new_rating();

Calculate new rating after the given games.

=head2 get_points_expected

  my $points_expected = $player->get_points_expected();

Calculate expected points according to rating differences between own rating
and opponents ratings.

=head2 get_performance

  my $performance = $player->get_performance();

Calculate performance according to average rating of opponents and percentage
score.


=head1 CONFIGURATION AND ENVIRONMENT

Games::Ratings requires no configuration files or environment variables.


=head1 DEPENDENCIES

This module relies on Games::Ratings which provides some generic methods, e.g.
  * new()
  * get_rating()
  * set_rating()
  * get_coefficient()
  * set_coefficient()
  * add_game()
  * remove_all_games()
  * DESTROY()


=head1 DIAGNOSTICS
 
At the moment, there are no error or warning messages that the module can
generate.

=for author to fill in:
    List every single error and warning message that the module can
    generate (even the ones that will "never happen"), with a full
    explanation of each problem, one or more likely causes, and any
    suggested remedies.

=over

=item C<< Error message here, perhaps with %s placeholders >>

[Description of error here]

=item C<< Another error message here >>

[Description of error here]

[Et cetera, et cetera]

=back


=head1 INCOMPATIBILITIES

None reported.


=head1 BUGS AND LIMITATIONS

No bugs have been reported.

At the moment, it's not possible to compute a FIDE rating for a previously unrated player.

Note, that a missing development coefficient (set via
$player->set_coefficient()) may lead to incorrect results. The program tries
to guess the correct factor according to the players rating, but it will err
for players new to the rating list with less than thirty games played.

Please report any bugs or feature requests to
C<bug-games-ratings@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Christian Bartolomaeus  C<< <bartolin@gmx.de> >>


=head1 ACKNOWLEDGMENTS

This module was inspired by Terrence Brannon's module Chess::Elo
(http://search.cpan.org/~tbone/Chess-Elo/).


=head1 SEE ALSO

http://en.wikipedia.org/wiki/Elo_rating for informations about the Elo system.

http://www.fide.com/info/handbook?id=11&view=category for informations about
the FIDE rating system (esp. point 10.0).


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, Christian Bartolomaeus C<< <bartolin@gmx.de> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
